ml-prims/src/matrix/gather.h

```diff
@@ -0,0 +1,295 @@
+
+  IndexT outRowStart = blockIdx.x * D;
+  MapValueT map_val = map[blockIdx.x];
+  StencilValueT stencil_val = stencil[blockIdx.x];
+
+  bool predicate = pred_op(stencil_val);
+  if(predicate){
+    IndexT inRowStart =  transform_op(map_val) * D;
+    for (int i = threadIdx.x; i < D; i += TPB) {
+      out[outRowStart + i] = in[inRowStart + i];
+    }
+	
+  }
+}
+
+/**
+ * @brief  gather conditionally copies rows from a source matrix into a destination matrix according to a transformed map.
+ *
+ * @tparam MatrixIteratorT      Random-access iterator type, for reading input matrix (may be a simple pointer type).
+ * @tparam MapIteratorT         Random-access iterator type, for reading input map (may be a simple pointer type).
+ * @tparam StencilIteratorT     Random-access iterator type, for reading input stencil (may be a simple pointer type).
+ * @tparam UnaryPredicateOp     Unary lambda expression or operator type, UnaryPredicateOp's result type must be convertible to bool type.
+ * @tparam MapTransformOp       Unary lambda expression or operator type, MapTransformOp's result type must be convertible to IndexT (= int) type.
+ *
+ * @param  in           Pointer to the input matrix (assumed to be row-major)
+ * @param  D            Leading dimension of the input matrix 'in', which in-case of row-major storage is the number of columns
+ * @param  N            Second dimension
+ * @param  map          Pointer to the input sequence of gather locations
+ * @param  stencil      Pointer to the input sequence of stencil or predicate values
+ * @param  map_length   The length of 'map' and 'stencil'
+ * @param  out          Pointer to the output matrix (assumed to be row-major)
+ * @param  pred_op      Predicate to apply to the stencil values
+ * @param  transform_op The transformation operation, transforms the map values to IndexT
+ * @param  stream       CUDA stream to launch kernels within
+ */
+template <typename MatrixIteratorT,
+	  typename MapIteratorT,
+	  typename StencilIteratorT,
+	  typename UnaryPredicateOp,
+	  typename MapTransformOp>
+void gatherImpl(MatrixIteratorT in,
+		int D,
+		int N,
+		MapIteratorT map,
+		StencilIteratorT stencil,
+		int map_length,
+		MatrixIteratorT out,		
+		UnaryPredicateOp pred_op,
+		MapTransformOp transform_op,
+		cudaStream_t stream){
+  // skip in case of 0 length input      
+  if(map_length <= 0 || N <= 0 || D <= 0)
+    return;
+
+  // signed integer type for indexing or global offsets
+  typedef int IndexT;
+
+  // map value type
+  typedef typename std::iterator_traits<MapIteratorT>::value_type MapValueT;
+
+  // map value type
+  typedef typename std::iterator_traits<StencilIteratorT>::value_type StencilValueT;
+      
+  // return type of MapTransformOp, must be convertable to IndexT
+  typedef typename std::result_of<decltype(transform_op)(MapValueT)>::type MapTransformOpReturnT;
+  ASSERT((std::is_convertible<MapTransformOpReturnT, IndexT>::value),
```

Comment:
this might as well be a `static_assert`, right?

Reply:
Yeah you are right, I have fixed it!